gui/src/app/init/analytics.js

```diff
@@ -1,75 +1,59 @@
-/* Copyright (C) 2017 Canonical Ltd. */
+/* Copyright (C) 2019 Canonical Ltd. */
 'use strict';
 
 
-/**
-  Factory that returns a function for sending analytics to GA.
 
-  @param {String} controllerAPI For checking status of the user.
-  @param {String} dataLayer The GA dataLayer instance.
-  @return {function} The curryed function for sending analaytics.
-*/
-const sendAnalyticsFactory = function(controllerAPI, dataLayer) {
-
-  /**
-    The retunrned function that sends stats to GA.
-
-    @param {String} category A sensible category name. Generally the
-      component name is a good start.
-    @param {String} action Some identifiable action.
-    @param {String} label Name the event. This might be enough for
-      some events.
-    @param {Object} value An optional single depth object for
-      extra information.
-  */
-  return function(category, action, label, value) {
-    // We want to check for required params to provide good feedback for
-    // developers - this is a fail fast way to ensure required fields are
-    // set. On the other hand, we don't want to block the execution in case
-    // of errors while sending analytics.
-    const requiredArgs = ['category', 'action', 'label'];
-    for (let i = 0, ii = requiredArgs.length; i < ii; i+= 1) {
-      if (!arguments[i]) {
-        console.error(`cannot send analytics: ${requiredArgs[i]} required`);
-        return;
-      }
-    }
-    if (!dataLayer) {
-      return null;
-    }
-    let loggedIn = undefined;
-    // Sometimes this doesn't get set...
-    // Always decorate with whether the user is logged in
-    if (controllerAPI) {
-      loggedIn = controllerAPI.userIsAuthenticated;
-    }
-    let valueObj = {
-      loggedIn: loggedIn
-    };
-    let valueArr = [];
-    let valueStr;
-
-    if (value) {
-      Object.keys(value).forEach(key => {
-        valueObj[key] = value[key];
-      });
+class Analytics {
+  constructor(data) {
+    if (data) {
+      this.category = data.category;
+      this.label = data.label;
+      this.getLabel = data.getLabel;
     }
-
-    Object.keys(valueObj).forEach(key => {
-      valueArr.push(`${key}:${valueObj[key]}`);
+  }
+
+  addComponent(component) {
+    let categories = this.category ? [this.category] : [];
+    categories.push(component._reactInternalFiber.elementType.name);
+    // Return a new analytics instance with the new data. This is so that we
+    // don't overwrite a top level instance with every fork as the analytics is
+    // passed into nested components.
+    return new Analytics({
+      category: categories.join(' > '),
```

Comment:
Maybe this should stay as an easier to manipulate data structure, like an array, until the event is actually sent. It would make it a little easier to tweak down the line if we wanted. And arguably a micro perf upgrade ;)

Reply:
I did have that, but I went for uniformity of flexibility as it didn't seem to give us much (just made the code more complicated).